### DISCUSSION TITLE
Help Ensure Sawppy V2 Components Are Available Worldwide

### DISCUSSION THREAD

**Discussion by Roger-random** [Roger-random/Sawppy_Rover#20] [Q&A]
Hello Sawppy builders,

I want to thank everyone who have built their own Sawappy and gave me feedback. I have since learned some Sawppy components can be difficult to get worldwide. I plan to redesign Sawppy version 2 to use more commonly available parts. I list some candidates below, and I would love feedback on this list from other rover builders. Especially those living outside of the United States.

- Is the candidate item easier to purchase than equivalent of Sawppy V1?
- Is it economical or is it overpriced?
- Do you have alternate suggestions?

[_**Structural Beams**_](https://newscrewdriver.com/2020/11/29/sawppy-issue-misumi-aluminum-extrusions/)

**Sawppy V1**: Misumi HFS3 15mm x 15mm aluminum extrusions. I knew Misumi had distributors worldwide, but I didn't know many of them only sold to businesses and not available to hobbyists. MakerBeam XL is an alternate, but at greater expense.

**Sawppy V2 Candidate**: 20mm x 20mm aluminum extrusions. Since they are used as structural members in 3D printers like the Ender 3 (& clones.) I believe they are more commonly available worldwide. True or false?

[**_Motors_**](https://newscrewdriver.com/2020/11/30/sawppy-issue-serial-bus-servo/)

**Sawppy V1**: LewanSoul LX-16A x 10 interfaced via Bus Linker board. These are not available worldwide, and only sporadically available within the United States. (See Issue #18)

**Sawppy V2 Candidates**

Steering:

Generic remote control (RC) servos. Available in a wide spectrum from budget low-end to powerful high end metal gear units. [Example on Amazon US](https://amzn.to/2IxgkZk)

Driving:

Brushed DC gearmotor.

- Budget: The ubiquitous yellow TT Gear Motors. [Example on Amazon US](https://amzn.to/39WO5i1) Problem: The motor may not be powerful enough, and the gears are plastic. But they are cheap!
- Midgrade: 25mm diameter gearmotor with simple gearbox. Optional encoder available. [Example on Amazon US](https://amzn.to/39U6Tyx) Problem: gear are metal but can still be fragile, sudden shock will [break tooth on output gear](https://newscrewdriver.com/2020/04/19/wheel-drive-motor-gearbox-swap-for-jpl-open-source-rover/).
- Durable: For better durability, Sawppy should use a planetary gearbox, which spreads shock load across multiple gears. But there doesn't seem to be a widely available commodity planetary gear motor? (Suggestions wanted)

Brushed DC Driver:

- Classic: L298N dual H-Bridge
- Newer: DRV8833
- Other?

NEMA 17 stepper motors.

Advantage: Precise motion control. Ball bearings typically built-in. Available worldwide(?)
Disadvantage: 3D printer NEMA 17s have little torque for rover wheel purposes. Higher torque (typically correlated to longer body) NEMA 17 motors exist but less common and steeply more expensive.

Stepper driver:

- 3D printer [stepper driver carriers](https://www.pololu.com/category/120/stepper-motor-drivers) are a popular form factor with wide variety of chips from classic A4988 to newer models like TMC2209.
- Discrete "TB6600" stepper driver modules (most of which aren't actually using a TB6600 chip anymore.)
- Other?

28BYJ stepper motors

Advantage: Very affordable, has a built-in gearbox and usually bundled with ULN2003 drivers. [Amazon US example](https://amzn.to/37PF7Ar)
Disadvantage: Not very powerful even with gearbox. Unipolar stepper motor design makes it incompatible with most stepper driver units. (Yes they can be hacked into bipolar steppers, but I'd like to make build instructions simple.)
Worldwide availability: Unknown.

Motor controller:

LX-16A had an integrated controller making wiring easy. Switching to servos + brushed DC/stepper motors makes wiring more complex and would also require a separate microcontroller to generate the necessary PWM control pulses. The popular Arduino Uno & Nano, unfortunately, do not have enough PWM outputs: We need to control 4 steering servos and 6 driving wheel motor controllers.
- Espressif ESP32 DevKit-C [Amazon US example](https://amzn.to/39Y853V)
- Multiple Arduinos. (1 for steering servos + 1 for rolling motors, etc.)
- Other candidates?

---

(*) Disclosure: All Amazon links in this message are affiliate links. Not because I'm trying to make money, but because it's a lot faster for me to get a short URL! Anyway, as an Amazon Associate I earn from qualifying purchases.

**Comment by srmainwaring** [comment]
Roger, for the motor controller can I suggest adding an option to use boards that can run Ardupilot. I'm in the process of getting my Sawppy to run on a Mateksys H743-Wing. To start with I'll use it with the LX16A serial bus board, but the board has 13 PWM outputs that could be used to run servos directly or command ESCs.  There are many alternative FCUs that support enough PWM channels to command the 10 motors needed. A FCU is more expensive than a nano or teensy but they come with integrated IMUs and other sensors (the GPS may be on-board or external), so the overall package starts to look competitive with buying separate components and integrating them yourself. You also get mission management, telemetry, ground station support and a host of other features, but I guess it depends on how much people want to build / code this for themselves as opposed to have something up and running.

**Reply by Roger-random** [reply]
I like the idea of supporting an upgrade path to Ardupilot, however I'm personally ignorant of that ecosystem and how to best support it. I'll need advice from people who know more.

I looked over the spec sheet for the Mateksys H743-Wing you mentioned. Hardware-wise it can certainly handle the PWM requirements. (I haven't researched the software side.) Not only does it have PWM, it also has UART, SPI, and I2C. Is this common of other Ardupilot flight control units available? If so it opens many options.

- FCU -- direct PWM control --> motor drivers

This would consume the majority of PWM expansion I/O. Are there typically plenty of open slots? Or would this be too greedy and prevent users from adding more modules? If the latter, then we'd need to use a coprocessor for PWM so only a single communication channel is consumed. But then the question is, what coprocessor, and what communication channel?

- FCU --(UART)--> ESP32 (for example) --(PWM)--> motor drivers
- FCU --(SPI)--> ESP32 --(PWM)--> motor drivers
- FCU --(I2C)--> ESP32 --(PWM)--> motor drivers

And there are probably other options I'm ignorant of.

- FCU --(other?)--> (?)

From an Ardupilot compatibility standpoint, what are the tradeoffs between these approaches, and which path would you suggest for Sawppy V2?

**Reply by srmainwaring** [reply]
There are many flight controllers to choose from, including both open and closed source hardware and ranging from cheaper hobby models to flight controllers suitable for commercial use. The Ardupilot documentation has this section on different options that are known to work: [Autopilot Hardware Options](https://ardupilot.org/rover/docs/common-autopilots.html). Because Sawppy would use scripting there are some additional requirements for the FCU outlined in this link about [Lua Scripts](https://ardupilot.org/copter/docs/common-lua-scripts.html). We could extract a sub-list of ones that would be suitable.

The usual set up is to run PWM to an ESC for each drive motor (your first config). So one configuration could be:

4 x (FCU > PWM > Steering Servo)
6 x (FCU > PWM > ESC > Drive Motor)

Taking the H743-Wing as an example that would leave 3 PWM channels to run servos for a camera gimbal or other peripherals.

There is support for different [ESC protocols](https://ardupilot.org/rover/docs/common-dshot.html#esc-related-configuration) , [SBus servos](https://ardupilot.org/rover/docs/common-servo-landingpage.html), so there's a good chance most typical use cases will be provided for.

I see it as an advantage that people can choose the motor and servo config that suits them and is available locally - provided the devices can be adapted to the frame and are correctly rated, the controller doesn't really care what's there as long as it can get a PWM signal to it. The rest is in the tuning (getting steering and throttle controllers tuned for a particular set up).

The main thing that is missing in Ardupilot is a rover frame matching Sawppy's 4 steering - 6 drive configuration, however that can be addressed by providing a custom Lua script to drive the PWM channels given the high level steering and throttle commands. I'm working on that at present and will post it when I've got it working and tested.

If you wish you can add an onboard companion computer to the FCU via a UART and then direct the FCU using Python scripts with [DroneKit](https://ardupilot.org/dev/docs/droneapi-tutorial.html) or even ROS - lots of choices! But you don't need all that to get a rover capable of running outdoor autonomous missions - the FCU, a RC unit, SiK radio telemetry link and laptop running a ground station will get you there.

**Reply by samuk** [reply]
"We need to control 4 steering servos and 6 driving wheel motor controllers. Espressif ESP32 DevKit-C Amazon US example."

One possible solution, which may not be the lowest cost, but which would add modularity would be to use the M5stack components;

- A core of;  ~ $4 [ESP32devc](https://www.aliexpress.com/item/4000090521976.html) & ~$5? [OpenCore](https://github.com/rosmo-robot/Open-Core-M5stack/blob/main/README.md)

Plus depending on user budget/preference one of;

- Lowest cost version +$17.99 [16 channel servo module](https://shop.m5stack.com/collections/stack-series/products/servo2-module-16-channels-13-2-pca9685) I'm unclear on power but should do 10x?

- Mid + $66 3x [2Amp DC & servo drivers stacked](https://shop.m5stack.com/collections/stack-series/products/goplus2-dc-motor-and-servo-driver-module-stm32f0) giving 6 x2amp drive motors plus servo control for steering.

- High + $90? - TBC 3x [Rosmo ESC](https://github.com/rosmo-robot/Rosmo_ESC) Untested but theoretically would drive 6x BLDC at 5amp. In addition you'd still need 4x lewansoul for drive.

I guess the main issue is that the 'open core' remains untested and unproven, and availability is uncertain. Our intention is to get it made and sold[ here](https://www.makerfabs.com/) with worldwide shipping, but appreciate that's a bit speculative at this point. Using the official M5 stack core would probably work out too expensive. 

On the upside, it would give users a lot of flexibility to experiment with different [modules](https://shop.m5stack.com/collections/stack-series?page=2) Open hardware purists would also have some options for modules as detailed[ here](https://oshwlab.com/Rosmo/m5-proto)


(moved to correct thread)

**Comment by nbering** [comment]
I took a few minutes to look for suppliers of 2020 aluminum extrusion, here in Canada. It does look like it's more accessible than the 15mm Misumi profile, but it's still kind of a specialty item. I had best success with 3D-printer suppliers, and Amazon.

If you wanted to look at materials that are _most_ available... angle or square tube profiles seem to be the universally accessible. You can get them at pretty much any hardware store, in a variety of materials including aluminum. They might be just a little harder to design for an easy build though, since you wouldn't have the pretty T-Slots to get good clamping force just about anywhere you want. You'd probably need to drill some holes in some cases.

**Reply by Roger-random** [reply]
Thank you for taking the time to do this.

It's disappointing that 20mm extrusions aren't much better than 15mm, but I appreciate the additional vote for metal tubes. I agree it would make Sawppy more difficult to build, but I also have other motivations for Sawppy evolution to require basic shop tools such as a drill press. If I go down that path, then metal tubes becomes a smaller additional step to take.

For aspiring rover builders without ready access to such tools, I plan to release a smaller simpler Sawppy that can be built using just hand tools.

The challenge with commodity metal tubes are dimensional tolerances far looser than extrusion beams. 20mm extrusion beams are usually quoted with tolerances < +/- 0.5mm. But I've found hardware store tubes to be up to 10% off nominal. (20mm +/- 2mm). It's a design challenge for sure, and I'm still thinking about it.

**Reply by nbering** [reply]
That's a great point. I hadn't thought about tolerances.

Maybe put one bolt through a hole in the tube, and another through threads in the 3d-printed part (tapped hole, captive nut, heat-set insert), to apply tension to the tube without actually going through it. Set-screws might be ideal for this since their points are usually designed to grip. The through-bolt would ensure the joint is secure (never slips out), and the tension screw would just keep it from wobbling.

Alternatively, your experimental work with crush-ribs might be useful here.

**Reply by Roger-random** [reply]
Related thought: if I could somehow design to accommodate such loose tolerances, there's another benefit. Such parts would accommodate both metric and imperial! For example: if the 3D printed parts could accommodate metal tubes of 25mm +/- 2.5mm, then it also easily accommodates 1 inch (~25.4mm)

I don't know if I can pull it off, but it'd be great if I could.

**Comment by samuk** [comment]
"It's disappointing that 20mm extrusions aren't much better than 15mm, "

I really like [20 x 20 Vslot ](https://openbuildspartstore.com/v-slot-20x20-linear-rail/)  It's available from:

[Openbuilds USA](https://openbuildspartstore.com/v-slot-20x20-linear-rail/) 

[Ratrig Portugal/EU](https://www.ratrig.com/aluminium-profiles/vslot.html)

[Aliexpress China](https://www.aliexpress.com/item/1005002679074398.html) £12.99 + £0.49 shipping to UK

[Banggood China](https://uk.banggood.com/Machifit-Black-100-1200mm-2020-T-slot-Aluminum-Extrusions-Aluminum-Profiles-Frame-for-CNC-Laser-Engraving-Machine-p-1484404.htm)

[Oozenest UK]( https://ooznest.co.uk/product/v-slot-linear-rail-20x20mm-cut-to-size/ ) 
With Royal mail shipping: Europe 3 – 10 Working days un/tracked service. 
Rest of World 5 – 14 Working days un/tracked service. 

Example Cost of 2x 500mm linear rail is £7.20 
plus
£4.49 shipping to Europe (Or £8.90 tracked)
£6.21 to rest of world; I  checked Argentina, Australia, Canada, Congo, Gambia, Guatemala (Or £11.21 tracked)

Payment methods are; Card payment, Paypal, Bank transfer, or purchase order.

**Reply by Roger-random** [reply]
These are very promising solutions. However, do note that worldwide shipping by the seller is only half the problem. The other half are import duties and customs requirements. Some countries have rules and/or fees so onerous they're only feasible for high-volume businesses and impractical for individual hobbyists.

**Reply by samuk** [reply]
I dropped Openbuilds an email to see how many countries they have distributors in.

**Reply by nbering** [reply]
> Some countries have rules and/or fees so onerous they're only feasible for high-volume businesses and impractical for individual hobbyists.

McMaster Carr wouldn't ship to me in Canada on a personal order. I could have just said I was a business, but I'm guessing it would have led to my arranging my own customs brokerage. I've been close to a project that had to do this for work before, and not a task I'd take on lightly.

McMaster's international shipping policy was pretty unusual in my experience. Misumi was easier to deal with, but I still had to pay pretty outrageous shipping rates. The shipping for the frame components cost at least as much as the product, if not more. I find this is pretty common when ordering over the border.

**Reply by samuk** [reply]
Openbuilds sent their [map of official distributors](https://www.google.com/maps/d/viewer?mid=1dDoqV3I3NL3RXQ4CiwV9VoR0N99L-zu3&ll=32.817073663940505%2C-36.02847408685306&z=2) TLDR: USA, Canada, Mexico, New Zealand & lot's in Europe.

**Reply by Roger-random** [reply]
More details on my comment: Sawppy's Hackaday.io page had interest from Argentina, where Misumi beams and LewanSoul servos were prohibitively expensive. They asked for (and got me thinking about) generic metal tubing and commodity NEMA17 stepper motors.

**Reply by samuk** [reply]
2x 500mm 20x20 vslot From Ooznest in the UK to Argentina would be £7.16 ($9.72 US)
Plus either
Royal Mail International Standard - Parcel: £6.21 ($8.73 US)
or
Royal Mail International Tracked - Parcel: £11.11 ($15.09 US)

Which doesn't seem too onerous?

**Reply by Roger-random** [reply]
As stated earlier: "do note that worldwide shipping by the seller is only half the problem. The other half are import duties and customs requirements. Some countries have rules and/or fees so onerous they're only feasible for high-volume businesses and impractical for individual hobbyists."

**Reply by samuk** [reply]
For Argentina, it looks like there'd be an additional [$5.54 US in duties & taxes](https://www.easyship.com/en-gb/duties-and-taxes-calculator/argentina?value=9.72&currency=USD&category=sport_leisure). That does start to add up I suppose.

I'll take a proper look at your BOM and do a calculation for the approximate real lengths.

**Reply by Roger-random** [reply]
Also, sometimes the customs duties can only be paid by a business registered as an importer with that country's government. An individual has to either find an existing import company willing to work on their behalf for a fee or they have to set up an import business themselves. (I do not expect people to start a company just for the purpose of importing Sawppy parts, though I would feel extremely honored if anyone does it.)

**Reply by samuk** [reply]
I wonder if it would be possible to design for 20x20mm square tube/profile, but remain agnostic as to whether it was a [generic aluminum square tube](https://www.google.com/search?q=20mm+aluminum+square+tube), Vslot or even [Timber](https://www.google.com/search?q=20mm+x+20mm+timber)?

**Reply by Roger-random** [reply]
That is my aspirational goal for Sawppy V2. It would be a huge challenge, but it would make the design much more adaptable.

**Reply by samuk** [reply]
3D printed structural beam holders which bolt together to clamp the beam, is that the basic idea?

**Reply by samuk** [reply]
I spotted these that work on a similar idea, not sure if they're a useful starting point? https://www.thingiverse.com/thing:5209449/files
